transformations/chinese_butter_fingers_perturbation/transformation.py

```diff
@@ -0,0 +1,227 @@
+                          chinese_words_database,
+                          prob,
+                          rare_word_prob,
+                          seed,
+                          max_outputs,
+                          consider_tone,
+                          word_level_perturb):
+    random.seed(seed)
+
+    perturbed_texts = []
+    for _ in itertools.repeat(None, max_outputs):
+        butter_text = ""
+        output = jieba.lcut(text)
+        if(word_level_perturb):
+            words_and_similar_word_dict_list = get_words_with_similar_pinyin(output,
+                                                                            rare_word_prob,
+                                                                            chinese_character_database,
+                                                                            common_chinese_character_database,
+                                                                            chinese_words_database,
+                                                                            consider_tone)
+            for dict in words_and_similar_word_dict_list:
+                perturb_word = dict['perturb_word']
+                similar_pinyins_words = dict['similar_pinyin_words']
+                if random.random() <= prob and len(similar_pinyins_words) != 0:
+
+                    new_chinese_character = random.choice(similar_pinyins_words)
+                else:
+                    new_chinese_character = perturb_word
+                butter_text += new_chinese_character
+        else:
+            for chinese_character in text:
+                similar_pinyins = get_characters_with_similar_pinyin(chinese_character,
+                                                                     rare_word_prob,
+                                                                     chinese_character_database,
+                                                                     common_chinese_character_database,
+                                                                     consider_tone)
+                if random.random() <= prob and similar_pinyins != "":
+
+                    new_chinese_character = random.choice(similar_pinyins)
+                else:
+                    new_chinese_character = chinese_character
+
+                butter_text += new_chinese_character
+        perturbed_texts.append(butter_text)
+    return perturbed_texts
+
+def get_characters_with_similar_pinyin(chinese_character,
+                                       rare_word_prob,
+                                       chinese_character_database,
+                                       common_chinese_character_database,
+                                       consider_tone):
+
+    pinyin_for_char_to_be_perturbed = ""
+    for char_dict in chinese_character_database:
+        if(chinese_character == char_dict['word']):
+            pinyin_for_char_to_be_perturbed = char_dict['pinyin']
```

Comment:
There can be more than one char with the same pinyin (with or without tones/accents) so I would need to loop through the character database to ensure that I retrieve all of them to be saved into the `pinyin_for_char_to_be_perturbed`

Reply:
Then I believe that there is an error here. It should be ```pinyin_for_char_to_be_perturbed += char_dict['pinyin']```.
Currently you only assign the last value found, hence my comment.

Reply:
Hi @mmeyer39 ,

You are right with your first comment! I am only getting the pinyin for a single character at that part of the code since I am passing each character individually here so there is only one result :D I've updated it with `break` as you suggested

I was thinking about what I was doing in `retrieve_from_database` when I saw your comment last time so I was thinking about the wrong thing.